auth/authlink/store.go

```diff
@@ -0,0 +1,240 @@
+package authlink
+
+import (
+	"context"
+	"crypto/rand"
+	"database/sql"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/jackc/pgtype"
+	uuid "github.com/satori/go.uuid"
+	"github.com/target/goalert/keyring"
+	"github.com/target/goalert/permission"
+	"github.com/target/goalert/util"
+)
+
+const (
+	MaxClaimQueryFrequency = 3 * time.Second
+	ExpirationDuration     = 10 * time.Minute
+)
+
+type Store struct {
+	db *sql.DB
+
+	unclaimed *sql.Stmt
+
+	create *sql.Stmt
+	claim  *sql.Stmt
+	verify *sql.Stmt
+	auth   *sql.Stmt
+
+	reset  *sql.Stmt
+	status *sql.Stmt
+
+	mx           sync.Mutex
+	closeCh      chan struct{}
+	waitRefresh  chan struct{}
+	startRefresh chan struct{}
+	wl           *Whitelist
+
+	keys keyring.Keyring
+}
+
+type Status struct {
+	ID string
+
+	ClaimCode  string
+	VerifyCode string
+
+	CreatedAt  time.Time
+	ExpiresAt  time.Time
+	ClaimedAt  time.Time
+	VerifiedAt time.Time
+	AuthedAt   time.Time
+}
+
+const codeAlphabet = "ABCDEFGHJKLMNPQRSTWXYZ23456789"
+
+func genCode(n int) (string, error) {
+	val := make([]byte, n)
+	for i := range val {
+		num, err := rand.Int(rand.Reader, big.NewInt(int64(len(codeAlphabet))))
+		if err != nil {
+			return "", err
+		}
+		val[i] = codeAlphabet[num.Int64()]
+	}
+
+	return string(val), nil
+}
```

Comment:
Like we talked about before briefly, could we separate this out to offer a digit only code? Also, I noticed that the `codeAlphabet` is missing `0` and `1`.

Perhaps something simple like this could work:
```
const codeAlphabet = "ABCDEFGHJKLMNPQRSTWXYZ"
const codeNums = "0123456789"

func genCode(n int, numOnly bool) (string, error) {
	var dict = codeAlphabet + codeNums
	if numOnly {
		dict = codeNums
	}
	val := make([]byte, n)
	for i := range val {
		num, err := rand.Int(rand.Reader, big.NewInt(int64(len(dict))))
		if err != nil {
			return "", err
		}
		val[i] = dict[num.Int64()]
	}

	return string(val), nil
}
```

Reply:
It's missing anything that could be confused -- e.g. 0 & O, l I & 1, u & v etc..

Instead of switching to numbers entirely, what do you think if we shorten it and maybe just offer confirmation? (i.e. no typing)